url8012c8d031_dmbb966_BBMS_tgz-pJ8-testbed_JNEATtestJ8/src/unit/JNEATIntegration.java

```diff
@@ -0,0 +1,747 @@
+					for (int j = 0; j < theList.size(); j++) {
+						Hex f = theList.elementAt(j);
+						viewage += (1.0 / (f.numSpots * power));
+					}
+				}
+				else {
+					viewage = theList.size();
+				}
+				
+				if (viewage > maxView) {
+					maxView = viewage;
+					destination = finger;
+				}
+			}
+			else i--;
+		}
+		
+		u.TeleportTo(destination);
+		if (GlobalFuncs.calcShared) {
+			u.emplaced = true;
+			u.UpdateSharedSpotting();
+		}
+		//GUI_NB.GCO("Placing " + u.callsign + " with maxView " + maxView);
+	}
+			
+	/** Based on sensor evaluations, determine where to deploy this unit */
+	public static void DeployOne(Unit u) {
+		if (u.org == null) {
+			GUI_NB.GCO("ERROR!  Unit " + u.callsign + " does not have an organism!");
+			return;
+		}
+		
+		if (u.orgType == OrganismTypeEnum.BASE_RANDOM) {
+			DeployRandomly(u);
+			return;
+		}
+		
+		if (u.orgType == OrganismTypeEnum.BASE_MAXHEX) {
+			DeployMaxView(u, 1);
+			return;
+		}
+		
+		boolean foundSpot = false;
+		int errCount = 0;
+		double resultThreshold = GlobalFuncs.networkResultThreshold;
+		int maxErrCount = 30;
+		
+		// Cycle through hex locations
+		while (!foundSpot) {
+			errCount++;
+					
+			Hex prospective = GlobalFuncs.scenMap.RandomHexReconZone();
+			
+			// Only chooses unoccupied hexes
+			if (prospective.HexUnit == null) {
+				// Uses the units' respective sensor function
+				double networkResult = u.UseSensors(prospective);
+												
+				//GUI_NB.GCO("Prospective hex: (" + prospective.x + ", " + prospective.y + ") is input: " + sensorInput + " with output: " + networkResult);
+				
+				if (networkResult > resultThreshold) {
+					//GUI_NB.GCO("Location accepted. Teleporting unit.");					
+					u.TeleportTo(prospective);
+					
+					if (GlobalFuncs.calcShared) {
+						u.emplaced = true;
+						u.UpdateSharedSpotting();
+					}					
+					
+					//u.DisplayLOSToRange(GlobalFuncs.visibility);
+					foundSpot = true;
+				}
+			}
+			
+			if (errCount > maxErrCount) {
+				resultThreshold -= 0.05;
+				errCount = 0;
+			}
+		}
+	}
+
+
+
+	/** Prints the population summary for each epoch to file */
+	public static void PrintSummaryOutput() {
+		FIO.appendFile(GlobalFuncs.summaryOutput, PrintSummaryLine());
+	}
+	
+	public static void PrintTestSummary() {
+		StringBuffer buf = new StringBuffer("");
+		
+		buf.append(GlobalFuncs.currentTestPath.toString() + ", ");
+		buf.append("COA " + GlobalFuncs.curCOA.name + ", ");
+		buf.append(GlobalFuncs.currentPopEpochNum + ", ");
+		buf.append((death_sum / death_count) + ", ");
+		buf.append((epoch_spotted / epoch_possibleSpot));
+		
+		FIO.appendFile(GlobalFuncs.summaryOutput, buf.toString());
+		
+	}
+	
+	public static String PrintTestSummaryKey() {
+		StringBuffer buf = new StringBuffer("");
+		
+		buf.append("Scenario, ");
+		buf.append("COA, ");		
+		buf.append("Generation, ");
+		buf.append("Avg Death Rate, ");
+		buf.append("Avg Team Performance");
+		
+		return buf.toString();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 356:364, 381:389.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.